feat(core): add `bootstrapApplication` function

AndrewKushnir opened the pull request:
Note: this is a draft as it relies on some missing functionality (such as a standalone injector service and other standalone components public API).

This commit implements the `bootstrapApplication` function that allows bootstrapping an application and pass a standalone component as a root component.


## PR Type
What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

pkozlowski-opensource commented on packages/platform-browser/src/browser.ts:
```diff
@@ -18,7 +18,55 @@ import {DomEventsPlugin} from './dom/events/dom_events';
+ *
+ * @publicApi
+ */
+export interface ApplicationConfig {
```
Question: is this interface something that can / should be shared across platforms? Would it makes sense to move it to core and export from here?

AndrewKushnir replied:
Currently the code is structured as follows:
- the `@angular/core` has an internal implementation of the `bootstrapApplication` function, which has more fields than a public version of the `bootstrapApplication` (for ex. it has a list of platform providers)
- the `@angular/platform-browser` uses an internal `bootstrapApplication` function and exposes a function with a slightly different interface (the public version) as a public API

My thinking is that the public function and supported interfaces should be co-located (i.e. located in `@angular/platform-browser`) as the `ApplicationConfig` might potentially vary depending on a platform (for ex. there might be more configuration required for SSR, etc).

alxhub replied:
We probably do want to share `ApplicationConfig` across platforms, but we'll probably do that if/when we move `bootstrapApplication` to core.

AndrewKushnir commented on packages/platform-browser/test/browser/bootstrap_spec.ts:
```diff
@@ -147,6 +149,20 @@ function bootstrap(
+      it('should work for simple components', async () => {
+        @Component({
+          selector: 'hello-app',
+          template: 'Hello from SimpleComp!',
```
Added extra checks into the code and corresponding tests.

AndrewKushnir committed:
feat(core): add the `bootstrapApplication` function

This commit implements the `bootstrapApplication` function that allows bootstrapping an application and pass a standalone component as a root component.

AndrewKushnir committed:
test(core): add integration test apps for the `bootstrapApplication` API

This commit adds 2 integration apps to verify the `bootstrapApplication` API behavior as well as keep track of the bundle size and retained symbols (tree-shaking).

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: public-api, integration-tests, fw-core

pkozlowski-opensource commented on packages/platform-browser/test/browser/bootstrap_spec.ts:
```diff
@@ -147,6 +149,187 @@ function bootstrap(
+        declarations: [NonStandaloneComp],
+      })
+      class NonStandaloneCompModule {
+      }
```
nit: personally I'm trying to avoid this style of tests where we declare upfront all the classes needed in subsequent tests. I know that many of the VE changes were written in this style but I find it hard to read and we were slowly moving away from this pattern - declaring classes needed by a particular test inside the corresponding `it(...)` block. 

Something we might want to discuss on the fwk / team sync.

AndrewKushnir replied:
Yeah, I was going back and forth with both styles and ended up moving Components outside of tests to make tests a bit more readable. But I don't have a strong opinion and happy to refactor in a followup PR if we decide to go different route.

pkozlowski-opensource commented on packages/platform-browser/test/browser/bootstrap_spec.ts:
```diff
@@ -147,6 +149,187 @@ function bootstrap(
+
+        // ApplicationRef should still allow bootstrapping non-standalone
+        // components into the same application.
+        appRef.bootstrap(NonStandaloneComp);
```
note to self: this API becomes _really_ confusing as we've made it OK to loose connection to the associated `NgModuleRef`...

pkozlowski-opensource commented on packages/platform-browser/test/browser/bootstrap_spec.ts:
```diff
@@ -147,6 +149,187 @@ function bootstrap(
+        expect(el2.innerText).toBe('Hello from Updated NonStandaloneComp!');
+      });
+
+      it('should throw when trying to use a non-standalone component', async () => {
```
nit: use => bootstrap?

pkozlowski-opensource commented on packages/platform-browser/test/browser/bootstrap_spec.ts:
```diff
@@ -147,6 +149,187 @@ function bootstrap(
+        }
+      });
+
+      it('should throw when trying to use a standalone directive', async () => {
```
nit: use => bootstrap?

pkozlowski-opensource commented on packages/platform-browser/test/browser/bootstrap_spec.ts:
```diff
@@ -147,6 +149,187 @@ function bootstrap(
+        }
+      });
+
+      it('should throw when trying to use a non-annotated class', async () => {
```
nit: use => bootstrap?

pkozlowski-opensource commented on packages/platform-browser/test/browser/bootstrap_spec.ts:
```diff
@@ -147,6 +149,187 @@ function bootstrap(
+              'NG0906: The StandaloneDirective is not an Angular component, ' +
+              'make sure it has the `@Component` decorator.';
+          expect((e as Error).toString()).toMatch(msg);
+        }
```
nit: wonder if those assertions could be rewritten using Jasmine's expect to throw, sth like:

```typescript
expect(() => {
    await bootstrapApplication(StandaloneDirective);
}).toThrowError('NG0906: The StandaloneDirective is not an Angular component, ' +
              'make sure it has the `@Component` decorator.';);
```

AndrewKushnir commented on packages/core/test/bundling/router/bundle.golden_symbols.json:
```diff
@@ -1190,6 +1190,9 @@
     "name": "getDeclarationTNode"
   },
+  {
+    "name": "getDirectiveDef"
```
The symbol is retained because of [packages/core/src/render3/definition.ts#L336](https://github.com/angular/angular/blob/master/packages/core/src/render3/definition.ts#L336), the `getDirectiveDef` is used inside of that function. I would expect the symbol to be there already in the golden file, but for some reasons it was not there before...

alxhub commented on packages/core/src/errors.ts:
```diff
@@ -63,7 +63,9 @@ export const enum RuntimeErrorCode {
   UNSAFE_VALUE_IN_RESOURCE_URL = 904,
-  UNSAFE_VALUE_IN_SCRIPT = 905
+  UNSAFE_VALUE_IN_SCRIPT = 905,
+  MISSING_GENERATED_DEF = 906,
```
I know these are internal names, but I think this error may be too vague. We could use `TYPE_IS_NOT_STANDALONE` in this case, with an error message saying basically "hey, this is not a standalone component as required by the API".

The fact that there's no Angular definition on the class is more of an implementation detail - the error as far as the user is concerned is that they needed to pass a standalone component, and didn't.

AndrewKushnir replied:
Thanks for the feedback Alex! I took a note on that and we'll do a cleanup once this PR lands (align error messages and error codes across the standalone part of the codebase).

alxhub commented on packages/core/src/application_ref.ts:
```diff
@@ -143,11 +149,107 @@ export function createPlatform(injector: Injector): PlatformRef {
+ *
+ * @returns A promise that returns an `ApplicationRef` instance once resolved.
+ */
+export function bootstrapApplication<RootComponentT>(config: {
```
There's no real purpose in having this function be generic - we can just use `Type<unknown>`.

alxhub commented on packages/core/src/application_ref.ts:
```diff
@@ -143,11 +149,107 @@ export function createPlatform(injector: Injector): PlatformRef {
+ * but avoid referencing `PlatformRef` class.
+ * This function is needed for bootstrapping a Standalone Component.
+ */
+export function createPlatformInjectorOnly(providers: StaticProvider[] = []): Injector {
```
```suggestion
export function createOrReusePlatformInjector(providers: StaticProvider[] = []): Injector {
```

alxhub commented on packages/core/src/application_ref.ts:
```diff
@@ -815,15 +924,20 @@ export class ApplicationRef {
     let componentFactory: ComponentFactory<C>;
-    if (componentOrFactory instanceof ComponentFactory) {
+    if (isComponentFactory) {
       componentFactory = componentOrFactory;
```
I'm surprised this assignment works. Without the explicit `instanceof` check, TS won't narrow the type of `componentOrFactory`.

AndrewKushnir replied:
That seems to be working fine (may be a recent change in TS?). Tested using [TS playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAECC0G9pgPwC5oQC4CcCWAdgOYDc0AvgFCiQwBCiy60BArgLYBGApjmVZQBmrAsCx4A9gWiCCACjA4iGWAB86ASkSVo0YFOzQ8EeAF5kSowWxhR3CYLgkdRx3OOwtCF7sVEAdGDQ5gDk2PjEIWQA9NFwgcGYuIRELuTQ3CAQ3Nq6vkoJ5gCMAAwxcXSFLBw8OGmUVEA) as well.

alxhub commented on packages/core/src/render3/errors.ts:
```diff
@@ -10,12 +10,30 @@
+  const componentDef = getComponentDef(type);
+  if (!componentDef) {
+    throw new RuntimeError(
+        RuntimeErrorCode.MISSING_GENERATED_DEF,
```
As mentioned before, we should use the same error code here, with a contextual message that either says "hey, this thing is not a standalone component" or "hey, this component is not standalone" depending on whether there's a component definition.

alxhub reviewed (approved):
Reviewed-for: public-api,size-tracking

alxhub commented on packages/platform-browser/src/browser.ts:
```diff
@@ -18,7 +18,53 @@ import {DomEventsPlugin} from './dom/events/dom_events';
+ *
+ * @publicApi
+ */
+export function bootstrapApplication<RootComponentT>(
```
Also no need to be have this be generic.

atscott reviewed (approved):
reviewed-for: public-api

AndrewKushnir commented:
Merge-assistance: PR is ready for merge.

atscott commented:
This PR was merged into the repository by commit 3e46a426c47bf6589952c7f2a17d914dd6f38dfe.
